### ISSUE TITLE
[PYD-144] Mypy plugin loses type information for classes subclassing GenericModel

### ISSUE THREAD

**Issue by twoolie** [pydantic/pydantic#5611]
The following code works fine with MyPy, but breaks when using the pydantic mypy plugin.

> ```
> from pydantic.generics import GenericModel
> from typing import Generic, TypeVar, Optional
> 
> T = TypeVar('T', int, str)
> 
> class HistoryField(GenericModel, Generic[T]):
>     value: Optional[T]
> 
> class DomainType(HistoryField[int]):
>     pass
> 
> thing = DomainType(value=None)
> var: Optional[int] = thing.value
> ```

`test.py:13: error: Incompatible types in assignment (expression has type "Optional[T]", variable has type "Optional[int]")  [assignment]`

It seems like the `DomainType`'s value field doesn't get the typevar updated with the `int` type.

Tested with Pydantic `1.9.2`, `1.10.7` and MyPy `0.990`, `1.1.0`, `1.2.0`

```
$ python -c "import pydantic.utils; print(pydantic.utils.version_info())"
             pydantic version: 1.9.2
            pydantic compiled: True
                 install path: /home/dev/.cache/pypoetry/virtualenvs/track-manager-_sAdQXr1-py3.10/lib/python3.10/site-packages/pydantic
               python version: 3.10.9 (main, Dec  6 2022, 18:44:57) [GCC 11.3.0]
                     platform: Linux-6.1.24-hardened1-x86_64-with-glibc2.35
     optional deps. installed: ['dotenv', 'email-validator', 'typing-extensions']
```

```
$ python -c "import pydantic.utils; print(pydantic.utils.version_info())"
             pydantic version: 1.10.7
            pydantic compiled: True
                 install path: /home/dev/.cache/pypoetry/virtualenvs/track-manager-_sAdQXr1-py3.10/lib/python3.10/site-packages/pydantic
               python version: 3.10.9 (main, Dec  6 2022, 18:44:57) [GCC 11.3.0]
                     platform: Linux-6.1.24-hardened1-x86_64-with-glibc2.35
     optional deps. installed: ['dotenv', 'email-validator', 'typing-extensions']
```

<sub>From [SyncLinear.com](https://synclinear.com) | [PYD-144](https://linear.app/pydantic/issue/PYD-144/mypy-plugin-loses-type-information-for-classes-subclassing)</sub>

**Comment by dmontagu** [comment]
I can confirm this is broken in v2 as well. If possible I will try to backport a fix to v1 when fixed for v2.

**Comment by dmontagu** [comment]
This is now fixed in v2.0 (which has just been released 🚀). Fixing it involved a significant update to the plugin, but if it is important to backport a fix to the v1 branch, it may be possible to do that without _too_ much pain. However, without significant interest specifically in supporting this in v1.10, I'm not currently planning to backport the fix (due to the complexity of the overhaul to the plugin).

However, if anyone wants to take a stab at backporting the fix, I would review a PR.
